modules/navier_stokes/src/fv/variables/INSFVVelocityVariable.C

```diff
@@ -0,0 +1,308 @@
+            {
+              auto & current_coeff = current_coeffs(i, j);
+              current_coeff = normal(i) * normal(j);
+              if (i == j)
+                current_coeff -= 1.;
+            }
+        }
+      }
+      else
+        // We are doing a one-term expansion for the extrapolated boundary faces, in which case
+        // we have no eqn. 2 and we have no second term in the LHS of eqn. 1. Instead we apply
+        // the element centroid value as the face value (one-term expansion) in the RHS of eqn.
+        // 1
+        grad_b += surface_vector * elem_value;
+    }
+    else
+      // We are not on an ebf, so we just use our getFaceValue method
+      grad_b += surface_vector * getFaceValue(neighbor, *fi, elem_value);
+
+    if (!volume_set)
+    {
+      // We use the FaceInfo volumes because those values have been pre-computed and cached.
+      // An explicit call to elem->volume() here would incur unnecessary expense
+      if (elem_has_info)
+      {
+        coordTransformFactor(
+            this->_subproblem, functor_elem.subdomain_id(), fi->elemCentroid(), coord);
+        volume = fi->elemVolume() * coord;
+      }
+      else
+      {
+        coordTransformFactor(
+            this->_subproblem, neighbor->subdomain_id(), fi->neighborCentroid(), coord);
+        volume = fi->neighborVolume() * coord;
+      }
+
+      volume_set = true;
+    }
+  };
+
+  Moose::FV::loopOverElemFaceInfo(*elem, this->_mesh, this->_subproblem, action_functor);
+
+  mooseAssert(volume_set && volume > 0, "We should have set the volume");
+  grad_b /= volume;
+
+  mooseAssert(ebf_faces.size() < UINT_MAX,
+              "You've created a mystical element that has more faces than can be held by unsigned "
+              "int. I applaud you.");
+  const auto num_ebfs = static_cast<unsigned int>(ebf_faces.size());
+
+  // test for simple case
+  if (num_ebfs == 0)
+    grad = grad_b;
+  else
+  {
+    // We have to solve a system
+    const unsigned int sys_dim =
+        lm_dim + num_ebfs + lm_dim * static_cast<unsigned int>(fdf_grad_centroid_coeffs.size());
+    DenseVector<ADReal> x(sys_dim), b(sys_dim);
+    DenseMatrix<ADReal> A(sys_dim, sys_dim);
+
+    // eqn. 1
+    for (const auto lm_dim_index : make_range(lm_dim))
+    {
+      // LHS term 1 coeffs
+      A(lm_dim_index, lm_dim_index) = 1;
+
+      // LHS term 2 coeffs
+      for (const auto ebf_index : make_range(num_ebfs))
+        A(lm_dim_index, lm_dim + ebf_index) = grad_ebf_coeffs[ebf_index](lm_dim_index) / volume;
+
+      // RHS
+      b(lm_dim_index) = grad_b(lm_dim_index);
+    }
+
+    unsigned int num_fdf_faces = 0;
+
+    // eqn. 2
+    for (const auto ebf_index : make_range(num_ebfs))
+    {
+      // LHS term 1 coeffs
+      A(lm_dim + ebf_index, lm_dim + ebf_index) = 1;
+
+      const bool fdf_face = ebf_faces[ebf_index].second;
+      const unsigned int starting_j_index = fdf_face ? lm_dim + num_ebfs + ebf_index * lm_dim : 0;
```

Comment:
I think it should be this?
```suggestion
const unsigned int starting_j_index = fdf_face ? lm_dim + num_ebfs + num_fdf_faces * lm_dim : 0;
```
We're only adding equation 3 for the fdf faces in my understanding

Reply:
You're exactly right